### PR TITLE
Add flag if task is speculative to TaskStatus

### PR DESCRIPTION
**Pull request by losipiuk**
On top of https://github.com/trinodb/trino/pull/17305

## Release notes

(x) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java [hunk 1]
```diff
@@ -491,8 +491,7 @@ public void failTaskRemotely(TaskId taskId, Throwable failureCause)
         private final SchedulingQueue schedulingQueue = new SchedulingQueue();
         private int nextSchedulingPriority;
 
-        private final Map<ScheduledTask, NodeLease> nodeAcquisitions = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (losipiuk) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java]: it was from "on top of" PR; that one was merged already

**Review by arhimondr — APPROVED** [review]

File: core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java [hunk 2]
```diff
@@ -1404,7 +1407,7 @@ public Optional<RemoteTask> schedule(int partitionId, ExchangeSinkInstanceHandle
                     splits,
                     noMoreSplits,
                     Optional.of(partition.getMemoryRequirements().getRequiredMemory()),
-                    false);  // TODO pass correct value here
+                    speculative);  // TODO pass correct value here
```
Comment (linzebing) [inline core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java:1410]: remove TODO

File: core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java [hunk 3]
```diff
@@ -2041,25 +2045,39 @@ public int getNonSpeculativeTaskCount()
...
+            if (previousTask != null) {
+                previousPrioritizedTask = getPrioritizedTask(previousTask);
+            }
+
+            if (!prioritizedTask.isSpeculative() && (previousPrioritizedTask == null || previousPrioritizedTask.isSpeculative())) {
+                // number of non-speculative tasks increased
                 nonSpeculativeTaskCount++;
             }
+
             // using negative priority here as will return entries with lowest pririty first and here we use bigger number for tasks with lower priority
```
Comment (linzebing) [inline core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java:2072]: pririty -> priority

**Review by linzebing — APPROVED** [review]

**Review by linzebing — APPROVED** [review]

File: core/trino-main/src/main/java/io/trino/execution/SqlTask.java [hunk 4]
```diff
@@ -88,6 +88,7 @@
     private final String taskInstanceId;
     private final URI location;
     private final String nodeId;
+    private final AtomicBoolean speculative = new AtomicBoolean(false);
```
[the remark this answers is not in the document]
Comment (losipiuk) [reply core/trino-main/src/main/java/io/trino/execution/SqlTask.java:91]: I will stick to `AtomicBoolean` to match what we use here with other fields

**Comment by losipiuk** [comment]
CI: https://github.com/trinodb/trino/issues/17158

**Pull request merged**